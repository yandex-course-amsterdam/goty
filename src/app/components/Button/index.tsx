import React, { ReactElement } from 'react'
import cn from 'classnames'

import style from './style.css'

type ButtonProps = {
  type: 'button' | 'submit'
  buttonText: string
  className?: string
  handleClick?: () => void
}

export const Button = ({ type, buttonText, className, handleClick }: ButtonProps): ReactElement => {
  return (
    <button
      onClick={handleClick}
      className={cn(style.button, className)}
      type={type === 'submit' ? 'submit' : 'button'}
      disabled
    >
      {buttonText}
    </button>
  )
}
