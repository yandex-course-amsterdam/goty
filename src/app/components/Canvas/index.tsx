import React, { useState, useRef, useEffect, ReactElement } from 'react'
import { State, Player, Enemy, Particle } from './entities'
import { fire } from './helpers/firemodes'

import style from './style.css'

export const Canvas = (): ReactElement => {
  const [state] = useState<State>(new State())
  const [score, setScore] = useState(0)
  const [enemiesSpawnInterval, setEnemiesSpawnInterval] = useState<ReturnType<typeof setInterval> | null>(null)
  const [showEndGamePopup, setShowEndGamePopup] = useState(false)
  const [ctx, setCtx] = useState<CanvasRenderingContext2D | null>(null)

  const animationFrameId = React.useRef(0)
  const boardRef = useRef<HTMLCanvasElement>(null)

  const spawnEnemies = (): void => {
    const interval = setInterval(() => {
      const player = state.getPlayer() as Player
      const enemy = new Enemy(
        { type: 'SMALL', killReward: 1, radius: 30, color: 'red', velocityMultiplier: 5 },
        ctx as CanvasRenderingContext2D,
        player
      )
      state.addEnemy(enemy)
    }, 750)

    setEnemiesSpawnInterval(interval)
  }

  const createPopup = () => {
    if (showEndGamePopup) return <div className={style.score}>We end {score}</div>
    return
  }

  const endGame = () => {
    setShowEndGamePopup(true)
    createPopup()

    clearInterval(enemiesSpawnInterval as ReturnType<typeof setInterval>)
    cancelAnimationFrame(animationFrameId.current)

    state.endGame()
  }

  const animate = (): void => {
    const { canvas } = ctx as CanvasRenderingContext2D
    const player = state.getPlayer() as Player
    const projectiles = state.getProjectiles()
    const enemies = state.getEnemies()
    const particles = state.getParticles()

    animationFrameId.current = requestAnimationFrame(animate)

    ctx!.fillStyle = 'white'
    ctx!.fillRect(0, 0, canvas.width, canvas.height)

    player.draw()

    projectiles.forEach((projectile, i) => {
      projectile.update()

      if (
        projectile.x + projectile.radius < 0 ||
        projectile.x - projectile.radius > canvas.width ||
        projectile.y + projectile.radius < 0 ||
        projectile.y - projectile.radius > canvas.height
      ) {
        setTimeout(() => {
          state.removeProjectile(i)
        })
      }
    })

    enemies.forEach((enemy, enemyIndex) => {
      enemy.update()

      const enemyPlayerDistance = Math.hypot(enemy.x - player.x, enemy.y - player.y)

      // end game
      if (enemyPlayerDistance - player.radius - enemy.radius < 1) {
        endGame()
      }

      projectiles.forEach((projectile, projectileIndex) => {
        const enemyProjectileDistance = Math.hypot(enemy.x - projectile.x, enemy.y - projectile.y)

        // projectile hit an enemy
        if (enemyProjectileDistance - projectile.radius - enemy.radius < 1) {
          for (let i = 0; i < enemy.radius * 0.5; i += 1) {
            const particle = new Particle(
              projectile.x,
              projectile.y,
              enemy.color,
              enemy.radius,
              ctx as CanvasRenderingContext2D
            )
            state.addParticle(particle)
          }

          const isDead = enemy.handleHit()

          setTimeout(() => {
            state.removeProjectile(projectileIndex)
          }, 0)

          if (isDead) {
            state.addScore(100)
            setScore(state.getScore())
            // state.addCredits(5)

            setTimeout(() => {
              state.removeEnemy(enemyIndex)
            }, 0)
          }
        }
      })
    })

    particles.forEach((particle, i) => {
      if (particle.alpha <= 0) {
        setTimeout(() => {
          state.removeParticle(i)
        }, 0)
      } else {
        particle.update()
      }
    })
  }

  const init = (): void => {
    const { canvas } = ctx as CanvasRenderingContext2D
    const player = new Player(canvas.width / 2, canvas.height / 2, 10, 'black', ctx as CanvasRenderingContext2D)
    state.registerPlayer(player)

    animate()
    spawnEnemies()
  }

  const handleClick = (evt: React.MouseEvent<HTMLCanvasElement>): void => {
    const player = state.getPlayer() as Player
    const angle = Math.atan2(evt.clientY - player.y, evt.clientX - player.x)

    const velocity = {
      x: Math.cos(angle) * 10,
      y: Math.sin(angle) * 10
    }

    fire(state, velocity, ctx as CanvasRenderingContext2D)
  }

  useEffect(() => {
    if (!boardRef.current) {
      return
    }

    setCtx(boardRef.current.getContext('2d'))

    if (ctx) {
      const canvasParent: HTMLElement = boardRef.current.parentElement as HTMLElement
      const canvasParentRect: DOMRect = canvasParent.getBoundingClientRect()
      const canvasWidth = canvasParentRect.width
      const canvasHeight = canvasParentRect.height
      boardRef.current.width = canvasWidth
      boardRef.current.height = canvasHeight

      init()
    }
  }, [ctx])

  return (
    <div className={style.canvasWrapper}>
      {createPopup()}
      <div className={style.score}>Score: {score}</div>
      <canvas id="board" ref={boardRef} onClick={handleClick} />
    </div>
  )
}
