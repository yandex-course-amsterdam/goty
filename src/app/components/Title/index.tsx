import React, { FC } from 'react'
import cn from 'classnames'

import style from './style.css'

interface IProps {
  title: string
  className?: string
}

export const Title: FC<IProps> = ({ title, className }): JSX.Element => (
  <h2 className={cn(style.title, className)}>{title}</h2>
)
