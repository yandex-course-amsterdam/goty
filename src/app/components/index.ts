export { Sidebar } from './Sidebar'
export { Avatar } from './Avatar'
export { Navigation } from './Navigation'
export { Main } from './Main'
export { Title } from './Title'
export { Description } from './Description'
export { Input } from './Input'
export { Textarea } from './Textarea'
export { Label } from './Label'
export { Error } from './Error'
export { ErrorBoundary } from './ErrorBoundary'
export { OfflineBar } from './OfflineBar'
export { Button } from './Button'
export { SubNavigation } from './SubNavigation'
export { Canvas } from './Canvas'
export { UserScore } from './UserScore'
export { Comment } from './Comment'
export { Comments } from './Comments'
export { Likes } from './Likes'
export { NewsList } from './NewsList'
export { SignUpForm } from './Forms/SignUpForm'
export { SignInForm } from './Forms/SignInForm'
export { FeedbackForm } from './Forms/FeedbackForm'
export { DetailsForm } from './Forms/DetailsForm'
export { PasswordForm } from './Forms/PasswordForm'
export { AvatarForm } from './Forms/AvatarForm'
export { Loader } from './Loader'
export { PrivateRoute } from './PrivateRoute'
export { ExitButton } from './ExitButton'
export { Authorization } from './Authorization'
export { Theme } from './Theme'
export { App } from './App'
