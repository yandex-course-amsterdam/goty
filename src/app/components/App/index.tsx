import React, { FC, useCallback, useEffect, useState } from 'react'
import { useSelector, useDispatch } from 'react-redux'

import { Switch, Route } from 'react-router-dom'
import { CSSTransition } from 'react-transition-group'

import {
  SignUpView,
  SignInView,
  ProfileView,
  GameView,
  ScoreView,
  NotFoundView
} from 'app/views'
import { OfflineBar, PrivateRoute, Authorization } from 'app/components'

import { postResult, getUserTheme } from 'app/api/Api'
import { fetchUserInfo } from 'app/actions'

import { StoreState } from 'app/reducers'
import { route } from 'app/enums'
import {
  getScore,
  removeScore,
  isServer,
  setUserTheme,
  storeUserTheme
} from 'app/utils'

import 'normalize.css'
import 'assets/main.css'
import '../../../fonts/fonts.css'

// startServiceWorker()

export const App: FC = (): JSX.Element => {
  const [isOffline, setIsOffline] = useState(isServer || !navigator.onLine)

  const dispatch = useDispatch()

  const loginStatus = useSelector(
    (state: StoreState) => state.loginStatus.status
  )

  const handleOfflineCallback = useCallback(() => setIsOffline(true), [])
  const handleOnlineCallback = useCallback(() => {
    setIsOffline(false)

    const storedScore = getScore()

    if (storedScore) {
      postResult(storedScore)
      removeScore()
    }
  }, [])

  useEffect(() => {
    const fetchUser = async () => {
      try {
        const data = await dispatch(fetchUserInfo())
        // Очередной странный тайпинг, учитывая возвращаемое по факту значение
        // Сейчас уже нет времени разбираться, буду потом сильнее вникать в ts
        // @ts-ignore
        const { data: theme } = await getUserTheme(data.id)
        setUserTheme(theme)
        storeUserTheme(theme)
      } catch (error) {
        console.warn(error)
      }
    }

    if (loginStatus) {
      fetchUser()
    }

    if (!isServer) {
      window.addEventListener('offline', handleOfflineCallback)
      window.addEventListener('online', handleOnlineCallback)
    }

    return () => {
      if (!isServer) {
        window.removeEventListener('offline', handleOfflineCallback)
        window.removeEventListener('online', handleOnlineCallback)
      }
    }
  }, [])

  return (
    <>
      <Switch>
        <Route exact path={route.auth}>
          <Authorization />
        </Route>
        <Route path={route.signUp}>
          <SignUpView />
        </Route>
        <Route path={route.signIn}>
          <SignInView />
        </Route>
        <PrivateRoute path={route.profile}>
          <ProfileView />
        </PrivateRoute>
        <PrivateRoute path={route.game}>
          <GameView />
        </PrivateRoute>
        <PrivateRoute path={route.score}>
          <ScoreView />
        </PrivateRoute>
        <Route path="*">
          <NotFoundView />
        </Route>
      </Switch>

      <CSSTransition
        in={isOffline}
        appear={isOffline}
        timeout={0}
        classNames="bar"
      >
        <OfflineBar />
      </CSSTransition>
    </>
  )
}
