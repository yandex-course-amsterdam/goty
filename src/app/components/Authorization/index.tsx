import React, { FC, useEffect, useState } from 'react'
import { useDispatch } from 'react-redux'
import { useLocation, Redirect } from 'react-router-dom'

import { authWithYandexOauth } from 'app/api/Api'
import { fetchUserInfo, setOAuthStatus } from 'app/actions'
import { LoaderView } from 'app/views'
import { route } from 'app/enums'

const useQuery = () => new URLSearchParams(useLocation().search)

export const Authorization: FC = (): JSX.Element => {
  const [isLoading, setIsLoading] = useState(true)
  const dispatch = useDispatch()

  const query = useQuery()
  const codeString = query.get('code')

  const getUserData = async () => {
    try {
      await dispatch(fetchUserInfo())
      setIsLoading(!isLoading)
    } catch (error) {
      console.error(error)
      setIsLoading(!isLoading)
    }
  }

  const signInWithOAuth = async () => {
    try {
      await authWithYandexOauth(codeString)
      await dispatch(fetchUserInfo())
      dispatch(setOAuthStatus(true))
      setIsLoading(!isLoading)
    } catch (error) {
      console.log(error)
      setIsLoading(!isLoading)
    }
  }

  useEffect(() => {
    if (codeString) {
      signInWithOAuth()
    } else {
      getUserData()
    }
  }, [])

  return isLoading ? <LoaderView /> : <Redirect to={route.game} />
}
