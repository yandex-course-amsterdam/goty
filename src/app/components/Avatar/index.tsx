import React, { FC } from 'react'
import { useSelector } from 'react-redux'
import cn from 'classnames'

import { StoreState } from 'app/reducers'

import style from './style.css'

interface IProps {
  className?: string
}

export const Avatar: FC<IProps> = ({ className }): JSX.Element => {
  const userInfo = useSelector((state: StoreState) => state.userInfo)

  const userAvatar = userInfo.avatar
    ? `https://ya-praktikum.tech/${userInfo.avatar}`
    : '/images/avatar.png'

  const avatarName = userInfo.displayName || 'Top game'

  return (
    <div className={cn(style.avatar, className)}>
      <img className={style.image} src={userAvatar} alt={avatarName} />
      <h1 className={style.title}>{avatarName}</h1>
    </div>
  )
}
