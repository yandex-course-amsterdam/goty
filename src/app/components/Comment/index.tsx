import React, { FC, useCallback } from 'react'
import { useSelector } from 'react-redux'

import { deleteComment } from 'app/api/Api'

import { CommentInterface } from 'app/interfaces'
import { StoreState } from 'app/reducers'

import style from './style.css'

interface IProps {
  comment: CommentInterface
  handleRemoveComment(id: number): void
}

export const Comment: FC<IProps> = ({
  comment: { id, text, user },
  handleRemoveComment
}) => {
  const userInfo = useSelector((state: StoreState) => state.userInfo)

  const removeComment = useCallback(
    async (commentId: number) => {
      await deleteComment(commentId)
      handleRemoveComment(commentId)
    },
    [handleRemoveComment]
  )

  const getUserAvatar = useCallback(
    (url: string | null) =>
      url ? `https://ya-praktikum.tech/${url}` : '/images/avatar.png',
    []
  )

  return (
    <div className={style.comment} key={id}>
      {userInfo.id === user.id && (
        <button
          type="button"
          className={style.commentDelete}
          onClick={() => {
            removeComment(id)
          }}
        >
          Delete comment
        </button>
      )}
      <div className={style.commentAvatar}>
        <img
          src={getUserAvatar(user.avatar)}
          alt={user.login || 'User avatar'}
        />
      </div>
      <div className={style.commentMain}>
        <div className={style.commentText}>
          {
            // unescape sanitized comments
            new DOMParser().parseFromString(text, 'text/html').documentElement
              .textContent
          }
        </div>
        <div className={style.commentMeta}>
          <div className={style.commentAuthor}>{user.login}</div>
          <div className={style.commentDate}>
            {new Date(user.createdAt).toLocaleString()}
          </div>
        </div>
      </div>
    </div>
  )
}
