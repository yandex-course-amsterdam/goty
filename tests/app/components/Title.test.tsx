import React from 'react'
import * as renderer from 'react-test-renderer'

import { Title } from 'app/components/Title'

describe('Title component', () => {
  test('Render itself correctly', () => {
    const component = renderer.create(<Title title="Oh hi" />).toJSON()
    expect(component).toMatchSnapshot()
  })

  test('Use className prop', () => {
    const component = renderer
      .create(<Title title="Oh hi" className="someClass" />)
      .toJSON()
    expect(component).toMatchSnapshot()
  })
})
