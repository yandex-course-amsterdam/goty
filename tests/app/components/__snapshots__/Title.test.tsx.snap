// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Title component Render itself correctly 1`] = `
<h2
  className=""
>
  Oh hi
</h2>
`;

exports[`Title component Use className prop 1`] = `
<h2
  className="someClass"
>
  Oh hi
</h2>
`;
